File: core/src/main/java/heast/core/network/pipeline/PacketEncryptor.java
package heast.core.network.pipeline;

import heast.core.logging.IO;
import heast.core.security.AES;
import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToByteEncoder;

import javax.crypto.SecretKey;

public final class PacketEncryptor extends MessageToByteEncoder<ByteBuf> {

    private final SecretKey key;

    public PacketEncryptor(SecretKey key) {
        this.key = key;
    }

    @Override
    protected void encode(ChannelHandlerContext ctx, ByteBuf msg, ByteBuf out) {
        var readable = msg.readableBytes();
        var bytes = new byte[readable];
        msg.readBytes(bytes);
        var encrypted = AES.encrypt(key, bytes);
        out.writeBytes(encrypted);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
        IO.error.println("Error while encrypting packet: " + cause.getMessage());
    }
}

File: Core/Network/Pipeline/PacketEncryptor.cs
using System.Security;
using System.Security.Cryptography;
using DotNetty.Buffers;
using DotNetty.Codecs;
using DotNetty.Transport.Channels;

namespace Core.Network.Pipeline; 

public class PacketEncryptor : MessageToByteEncoder<IByteBuffer> {
    
    private Aes KeyPair { get; }
    
    public PacketEncryptor(Aes keyPair) {
        KeyPair = keyPair;
    }

    protected override void Encode(IChannelHandlerContext context, IByteBuffer message, IByteBuffer output) {
        var rb = message.ReadableBytes;
        var bytes = new byte[rb];
        message.ReadBytes(bytes);
        
        Span<byte> encrypted = new();
        var success = KeyPair.TryEncryptCfb(bytes, KeyPair.IV, encrypted, out _, PaddingMode.PKCS7);
        if (!success) throw new SecurityException("Failed to encrypt packet!");
        output.WriteBytes(encrypted.ToArray());
    }

    public override void ExceptionCaught(IChannelHandlerContext context, Exception exception) {
        context.DisconnectAsync();
        Console.WriteLine($"Client kicked: {exception.Message}");
    }
}